api/src/test/java/org/pmiops/workbench/access/UserServiceAccessTest.java

```diff
@@ -170,131 +178,380 @@ public void test_updateUserWithRetries_register_includes_others() {
+          return userDao.save(user);
+        });
+  }
+
+  @Test
+  public void test_updateUserWithRetries_dua_unbypassed_aar_expired_noncompliant() {
+    providedWorkbenchConfig.access.enableDataUseAgreement = true;
+    providedWorkbenchConfig.access.enableAccessRenewal = true;
+
+    final long expirationWindow = 365L;
+    providedWorkbenchConfig.accessRenewal.expiryDays = expirationWindow;
+    final Timestamp expired =
+        Timestamp.from(START_INSTANT.minus(expirationWindow + 1, ChronoUnit.DAYS));
+
+    assertUnregisters(
+        user -> {
+          user.setDataUseAgreementBypassTime(null);
+          user.setDataUseAgreementCompletionTime(expired);
+          user.setDataUseAgreementSignedVersion(userService.getCurrentDuccVersion());
+          return userDao.save(user);
+        });
+  }
+
+  // Publications confirmation is subject to annual renewal and cannot be bypassed.
+
+  @Test
+  public void test_updateUserWithRetries_publications_not_confirmed() {
+    providedWorkbenchConfig.access.enableAccessRenewal = true;
+    providedWorkbenchConfig.accessRenewal.expiryDays = 365L;
+    assertUnregisters(
+        user -> {
+          user.setPublicationsLastConfirmedTime(null);
+          return userDao.save(user);
+        });
+  }
+
+  @Test
+  public void test_updateUserWithRetries_publications_expired() {
+    providedWorkbenchConfig.access.enableAccessRenewal = true;
+
+    final long expirationWindow = 365L;
+    providedWorkbenchConfig.accessRenewal.expiryDays = expirationWindow;
+    final Timestamp expired =
+        Timestamp.from(START_INSTANT.minus(expirationWindow + 1, ChronoUnit.DAYS));
+
+    assertUnregisters(
+        user -> {
+          user.setPublicationsLastConfirmedTime(expired);
+          return userDao.save(user);
+        });
+  }
+
+  // Profile confirmation is subject to annual renewal and cannot be bypassed.
+
+  @Test
+  public void test_updateUserWithRetries_profile_not_confirmed() {
+    providedWorkbenchConfig.access.enableAccessRenewal = true;
+    providedWorkbenchConfig.accessRenewal.expiryDays = 365L;
+    assertUnregisters(
+        user -> {
+          user.setProfileLastConfirmedTime(null);
+          return userDao.save(user);
+        });
+  }
+
+  @Test
+  public void test_updateUserWithRetries_profile_expired() {
+    providedWorkbenchConfig.access.enableAccessRenewal = true;
+
+    final long expirationWindow = 365L;
+    providedWorkbenchConfig.accessRenewal.expiryDays = expirationWindow;
+    final Timestamp expired =
+        Timestamp.from(START_INSTANT.minus(expirationWindow + 1, ChronoUnit.DAYS));
+
+    assertUnregisters(
+        user -> {
+          user.setProfileLastConfirmedTime(expired);
+          return userDao.save(user);
+        });
+  }
+
+  // checks which power most of these tests - confirm that the unregisteringFunction does that
+  private void assertUnregisters(Function<DbUser, DbUser> unregisteringFunction) {
```

Comment:
I think this pattern is pretty good, and it eliminates a lot of boilerplate from the test code, but I suggest renaming the method - I think it's too surprising that a method `assertX` has a side-effect of first registering the user, before applying the callback. I'd expect an assert method to be free of side effects

Maybe something like `testUnregistration()`, `doTestUnregister()`, `verifyUnregister()`

Reply:
Well now Codacy is unhappy.  heh.

Reply:
I vote to disable this pattern, it seems too limiting. Feel free to turn off the pattern in Codacy if you agree (or let me know if you have any issues accessing the style patterns).

Reply:
The AoU org appears to be out of Codacy seats, so I can't do this.  #5119 is a silly PR to trigger the rule.